Tool to unroll all repeats. Fix #283592

velochy opened the pull request:
Feature to unroll all repeats (prev: linearize). 

In active use (with MatchMySound - we have a custom build), but seems like it would be useful to other users too (#283592).

The core idea of the algorithm is quite simple:
A) Figure out repeated sections (this is provided by repeatsList used in playback)
B) Clone each repeated part at the end of the current score measure-by-measure, one repeated section at a time
C) Do a pass over the final result, removing all repeat markings (repeat barlines, seignos, codas, fines, etc).

To avoid messing up the beginning (which often has extra formatting), we currently clone the entire score and skip copying the first repeat segment, instead cutting the clone down to the right length.

The current solution handles parts rather badly, as it destroys and then re-creates them. This is mainly because it is very hard to keep all the parts synchronized when measures are being copied so it is far easier to do it only in the master score and then have the parts re-calculated.

EDIT: I consider this finished

Jojo-Schmitz commented on mscore/linearize.cpp:
```diff
@@ -0,0 +1,343 @@
+//  MuseScore
+//  Music Composition & Notation
+//
+//  Copyright (C) 2002-2014 Werner Schweer
```
2019 I guess, on a new file like this?

velochy replied:
Fixed

Jojo-Schmitz commented on mscore/linearize.cpp:
```diff
@@ -0,0 +1,343 @@
+namespace Ms {
+
+   // Influenced strongly by appendScore in score.cpp
+   Segment * appendCopiesOfMeasures(Score * score,Measure * fm,Measure * lm, Segment * cur_time_sigs) {
```
No indent here on this level

velochy replied:
Noted and fixed

Jojo-Schmitz commented on mscore/linearize.cpp:
```diff
@@ -0,0 +1,343 @@
+      TieMap tieMap;
+      MeasureBase * cmb = toMeasureBase(fm);
+      while (true) {
+         MeasureBase * nmb;
```
but otherwose 6 spaces pleas, not just 4

velochy replied:
@Jojo-Schmitz is there a formatting description file somewhere for MuseScore? I did as much as I could by hand but the easiest way to make sure nothing was missed would be to run an automated formatter.

Jojo-Schmitz replied:
Yes, there is, see https://musescore.org/en/handbook/developers-handbook/finding-your-way-around/musescore-coding-rules

velochy replied:
@Jojo-Schmitz ok, thank you. After reading the coding rules I see that I clearly violate a few more of them. I will try to fire up QTCreator and see how much it can fix as opposed to what still needs to be done by hand.

Jojo-Schmitz replied:
for QtCreator there an xml file at the end of that handbook pages that you can download and import in QtCreator for it to obey most rules automagically

velochy replied:
@Jojo-Schmitz I have opened the file in QtCreator and told it to use the xml style file. It is not auto-fixing the mistakes, or even pointing them out, however. And when googling, it seems you need to download a plugin to beautify the code and the plugin uses either clang or artistic syntax, not the xml you provided.

Sorry if this is a stupid question, but I usually use another code environment so qtcreator is new for me.

anatoly-os replied:
It is possible to compile MuseScore using Visual Studio if this is the one you are get used to. And XCode ofc.

velochy replied:
@anatoly-os no, I'm on Linux (Ubuntu 18.04 LTS). I am just used to using sublime for editing as all my other work is not in Qt. 
It may make sense to invest a bit of time to create an equivalent .clang-format file as that is more universal, being supported by most editors (including QtCreator)

velochy commented:
I have just signed the CLA as "Margus Niitsoo"

drowo commented:
Very much appreciated feature! I always use jumps and repeates during "protoyping" of a song and after a while have linearize before I start with the arrangement.  
Question: does it modify the current score, or does it actually create a new score? The latter would be safer.

velochy commented:
It can do both in theory, although we use the "new score" option in deployment, meaning the in-place option is likely to be untested.

drowo commented:
I've created a feature request in the issue tracker.
https://musescore.org/en/node/283592

Jojo-Schmitz commented:
Now mention "fix #283592" in the commit and PR title, so the issue gets auto-closed on merge

ChristianRomberg commented:
Hi,
playing with this a little bit, it seems that cloneMeasure has to be modified, because the number of staves of cloned measures is twice of what it should be (which is equal to the number of staves of the score). The issue (https://musescore.org/en/node/100246) was reported on the issue tracker a long time ago, and I think this might be a good opportunity to fix it. Thanks for your work!

velochy commented:
I spent some time trying to get this to a better shape.
a) Moved it to the tools menu, which makes more sense
b) Added code for the test as well as a rather complex test case (which currently clearly fails)
c) Moved to checking for underlying change events to decide if a key/clef/timesig is from copying the first measure of the staff (as it then has the staff beginning elements which get copied too)
d) Made some progress with real key, clef and tempo signature changes, but there is still much to do.

I plan to continue tomorrow. Any comments would be greatly appreciated.

Jojo-Schmitz commented:
Bad merge commit, you rebased wrongly

velochy commented:
> Bad merge commit, you rebased wrongly. Better use something like `git pull --rebase upstream master`
Thank you. Rebased the new commits with a force push for now and will do rebase pulls going forward

velochy commented:
The current version seems to handle key, tempo and clef changes correctly even in the rather complex test case I added as the automatic test.

However, for some reason, when linearize is called in the UI, key and time signature changes often get repeated twice for some reason and I can not figure out how or why this happens.

@anatoly-os maybe you can help?

ChristianRomberg commented:
I think those are courtesy sigs. They Display the same as regulär keysig Elements, and they are displayed when the keysig is in the first measure of a system

> Am 15.02.2019 um 03:59 schrieb Margus Niitsoo <notifications@github.com>:
> 
> The current version seems to handle key, tempo and clef changes correctly even in the rather complex test case I added as the automatic test.
> 
> However, for some reason, when linearize is called in the UI, key and time signature changes often get repeated twice for some reason and I can not figure out how or why this happens.
> 
> @anatoly-os maybe you can help?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

velochy commented:
That sounds quite likely. Do you have ideas regarding how to best deal with them?

Jojo-Schmitz commented:
It is a different bug, dealt with in #4656

velochy commented:
Unfortunately, pulling in the commit for that issue does not fix the doubling here, so it is either something else or the fix needs further work. Would be happy to assist in any way I can

velochy commented:
Empirically: filtering out all segments of type KeySigAnnounce and TimeSigAnnounce seems to achieve the desired result. Could anyone explain what these are and where/how they may get added?

ChristianRomberg commented:
I don't know where the code is in for version 3, but read206 adds KeySigAnnounces when the KeySig would be at the end of a measure.
https://github.com/musescore/MuseScore/blob/master/libmscore/read206.cpp#L3087-L3093

Jojo-Schmitz commented:
Code for MuseScore 3 should be in read301.cpp

ChristianRomberg commented:
One thing that would also be cool is handling lyrics. But I guess that would require an awful lot of contextual information 

> Am 15.02.2019 um 16:22 schrieb Joachim Schmitz <notifications@github.com>:
> 
> Code for MuseScore 3 should be in read301.cpp
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

velochy commented:
@ChristianRomberg indeed - lyrics would be great, but would likely require a fair bit of tinkering to get to wrok. I believe it is best to get this done without first, and then add the lyrics support down the line.

velochy commented:
@Jojo-Schmitz @anatoly-os 
I am now mostly happy with the PR as it stands and would ask you to do a more thorough review.

Getting it to work required modifying cloneMeasure to take an extra parameter (tick), which should be fine considering the function is only used in appendScore and I fixed it to use the correct value there. 
cloneMeasure also had a few bugs that needed to be fixed to get this to work (including #100246 where I used the fix originally proposed by @ericfont although I believe @mattmcclinch was not quite happy with that and was looking into it a bit deeper).

As recommended in the chat by @mattmcclinch I removed all calls to doLayout() and fixTicks() except the one at the very end, and so far everything seems to be working, so I assume it was safe to do so. 

I have added an automated test case that has a rather difficult combination of clef/key/time signature changes as well as a few slurs/ties and an excerpted part and it seems to handle the linearization on it as expected, giving me confidence this feature is in fact rather robust as it is.

Please let me know what else needs doing for this to be considered for merging into master

velochy commented:
Also: I am open to a better name than "Linearize" as it is not immediately clear what it means and it may be easily confused with linear layout.
"Unroll repeats" maybe? In this case, should the main file name be unrollrepeats.cpp?

mattmcclinch commented on libmscore/measure.cpp:
```diff
@@ -2792,15 +2792,14 @@ Fraction Measure::stretchedLen(Staff* staff) const
 
-      for (MStaff* ms : _mstaves)
-            m->_mstaves.push_back(new MStaff(*ms));
+      Q_ASSERT(sc->staves().size() >= _mstaves.size()); // destination score must we're cloning into have at least as many staves as measure being cloned
```
This comparison causes a compiler warning because the size() functions for QList and std::vector return different types. Also, the word "must" in the comment is in the wrong place. I would change it to
```
      Q_ASSERT(sc->staves().size() >= int(_mstaves.size())); // destination score we're cloning into must have at least as many staves as measure being cloned
```

Other than that, I am fine with this.

As for the original for loop, it was incorrect because it added MStaff objects based on the old measure's score to the new measure's _mstaves list. And the Measure constructor already adds MStaff objects based on the correct score to the measure's _mstaves list.

velochy replied:
@mattmcclinch fixed this. Thank you for helping me look into this yesterday.

mattmcclinch commented on libmscore/measure.cpp:
```diff
@@ -2799,8 +2799,8 @@ Measure* Measure::cloneMeasure(Score* sc, int tick, TieMap* tieMap)
-      Q_ASSERT(sc->staves().size() >= _mstaves.size()); // destination score must we're cloning into have at least as many staves as measure being cloned
-
+       Q_ASSERT(sc->staves().size() >= int(_mstaves.size())); // destination score we're cloning into must have at least as many staves as measure being cloned
+       
```
Sorry to do this, but now there is an extra space in front of Q_ASSERT, and the next line has a bunch of unnecessary spaces as well.

Also, are you planning on squashing your commits? Are you waiting for more review first?

velochy replied:
@mattmcclinch I gave this to a friend to test and he discovered one more issue I'm trying to take care of: pickup measures that if at the beginning of repeat should be merged to the last measure of the previous repeat section.

Will fix the Q_ASSERT in the same commit and then squash all of them for easier review.

velochy commented:
Fixed a few bugs and added one more test case to check the pickup measure merge. And squashed all the commits as you suggested.

PS: @mattmcclinch thank you for the help with finding cmdJoinMeasures

velochy commented on libmscore/unrollrepeats.cpp:
```diff
@@ -0,0 +1,352 @@
+      for (Excerpt * e : ((MasterScore*)score)->excerpts()) {
+            excerpts.append(e);
+            qDebug() << "Removing ", e->title();
+            score->masterScore()->deleteExcerpt(e);
```
Don't really get how that would work. The excerpts need to be re-generated but copying them would not achieve that. Or would it?

dmitrio95 replied:
No, copying itself wouldn't, but you could add that copy to `excerpts` list and use it as you do it right now. I just point that using the original `Excerpt` object for that purpose may be potentially dangerous.

velochy replied:
@dmitrio95 I am honestly admitting I do not understand the partScores and excerpts well enough to know if what I am doing is correct or not, but I changed
excerpts.append(e) to excerpts.append(e->clone());
which, if I understood you correctly, should take care of this issue for now.
Please do let me know if that is not quite what you meant, or if you believe other steps need to be taken.

velochy replied:
@dmitrio95  ... or I would have if it had a copy constructor or a clone function, but it does not...
So for now, I replaced it with
`            Excerpt * ne = new Excerpt(score);
            ne->setTitle(e->title());
            ne->setParts(e->parts());`
Which seems to be sufficient for the current purposes.
I could go in and add a copy constructor to Excerpt too, but in this case, more fields likely need to be copied for it to work in other use cases, and since I really do not understand how that part of the code works, it would be rather dangerous for me to just leave it there...
@dmitrio95 @anatoly-os how would you recommend I proceed?

dmitrio95 replied:
Copying tracks mapping (`e->tracks()`) would also be necessary to restore correctly those excerpts which have separate voices extracted rather than whole parts. Other than that, using a copying constructor would certainly seem to be a better solution but defining a proper copy constructor for `Excerpt` would require copying a `partScore` which is redundant here (and may be really slow). Maybe a compromise could be defining a copy constructor with some boolean `copyPartScore` parameter, something like:
```
Excerpt(const Excerpt& ex, bool copyPartScore = true)
   : _oscore(ex._oscore), _title(ex._title), _parts(ex._parts), _tracks(ex._tracks)
      {
      _partScore = (copyPartScore && ex._partScore) ? ex._partScore->clone() : nullptr;
      }
```
Then it would be possible both not to worry about `Excerpt`'s internals in this code and avoid the unnecessary score copying.

velochy replied:
@dmitrio95 added the constructor and used it in my code with copyPartScore set to false as you recommended.

velochy commented on libmscore/unrollrepeats.cpp:
```diff
@@ -0,0 +1,352 @@
+            ns->setTick2(spanner->tick2() - btick + tickOfAppend);
+            if (ns->isSlur()) {
+                  // set start/end element for slur
+                  ns->setStartElement(0);
```
@dmitrio95 copy-pasted from appendScore in score.cpp . I don't really understand your proposal though, so maybe you can elaborate a bit?.

dmitrio95 replied:
Ah, that code is from another part of libmscore. Then it is interesting why it is done so here.
I mean, if you just write
```
ns->computeStartElement();
ns->computeEndElement();
```
then there would probably no need for all that code below to set start and end element properly. In fact, I am not even sure that start and end elements are needed at all at this stage, perhaps only resetting them to `nullptr` would be enough.

And it is very curious that this all is done only for slurs while other spanners can also have start/end elements, it would be logical to do that for all spanners instead.

So I would propose to check whether replacing all that `if` block with
```
ns->setStartElement(nullptr);
ns->setEndElement(nullptr);
// or, alternatively
ns->computeStartElement();
ns->computeEndElement();
```
would work properly. If so I believe it would be preferable to use one of these options rather than preserving this chunk of code.

velochy replied:
@dmitrio95 This seems to have worked. Thank you for suggesting this.

velochy commented:
Rebased to latest master which meant moving to fractional ticks. Tests seem to pass though so it seems ok. 

Also: moved appendMeasuresFromScore to Measure and changed appendScore to use it (as this is where I originally got the code for it anyways so this avoids code duplication).

velochy commented:
@dmitrio95 @Jojo-Schmitz @anatoly-os I have implemented all the style fixes as well as the more substantial recommendations from dmitri, and consider this feature ready. Maybe mark it "review" instead of "work in progress" now?

JoshuaBonn1 commented on libmscore/score.cpp:
```diff
@@ -2067,59 +2034,132 @@ bool Score::appendScore(Score* score, bool addPageBreak, bool addSectionBreak)
+            
+            // check if time signature needs to be changed
+            if (!ostaff->timeSig(otick) || !staff->timeSig(ctick)) {} // yes, this can happen.
+            else if (!((*ostaff->timeSig(otick)) == (*staff->timeSig(ctick)))) {
```
Rather than this slightly confusing code, I think you should just make two variables and only one if statement. That way it’ll be immediately obvious as to what you are trying to do. Also, it will simplify some of the code later on.

velochy replied:
@JoshuaBonn1 I do not quite understand what you mean by "two variables" here. Could you elaborate a bit more?

JoshuaBonn1 replied:
I realize that I made a mistake and I see what you are doing. Now it's on to personal preference of reducing the code a bit. However, unless it breaks something, I think the extra parenthesis around the `*ostaff->timeSig(otick)` and `*staff->timeSig(ctick)` are excessive.

What follows below is simply preference. You may ignore it.

`ostaff->timeSig(otick)` and `staff->timeSig(ctick)` can be made into two variables.
`TimeSig* ts = staff->timeSig(ctick);`
`TimeSig* ots = ostaff->timeSig(otick);`
`if (!ts || !ots) {}`
Then replace with `ts` and `ots` where possible later on in the code.

mattmcclinch replied:
Perhaps this is what Joshua was suggesting:
```
TimeSig* ts1 = ostaff->timeSig(otick);
TimeSig* ts2 = staff->timeSig(ctick);
if (ts1 && ts2 && (ts1->sig() != ts2->sig()) { 
```
I presume that it is not necessary to test for full equivalence between the two TimeSig objects, and that comparing their sig() values is enough.

JoshuaBonn1 commented on libmscore/unrollrepeats.cpp:
```diff
@@ -0,0 +1,207 @@
+                  }
+            else if (e->isBarLine()) {
+                  toBarLine(e)->setBarLineType(BarLineType::NORMAL);
+                  }
```
The for loop should have brackets. The inner if statement does not need them.

JoshuaBonn1 commented on libmscore/score.cpp:
```diff
@@ -2067,59 +2034,132 @@ bool Score::appendScore(Score* score, bool addPageBreak, bool addSectionBreak)
+            else if (staff->currentKeyTick(ctick) == ctick &&
+                     staff->key(ctick - Fraction::fromTicks(1)) == ostaff->key(otick)) {
+                        Segment* ns = firstAppendedMeasure->first(SegmentType::KeySig);
+                        if (ns) ns->remove(ns->element(trackIdx));
```
new line for body

velochy replied:
@JoshuaBonn1 I do not understand what you mean here.

JoshuaBonn1 commented on libmscore/score.cpp:
```diff
@@ -2067,59 +2034,132 @@ bool Score::appendScore(Score* score, bool addPageBreak, bool addSectionBreak)
+                  nsig->setScore(this);
+                  nsig->setTrack(trackIdx);
+                  ns->add(nsig);
+                  //undoAddElement(nsig);
```
unnecessary?

JoshuaBonn1 commented on libmscore/score.cpp:
```diff
@@ -2067,59 +2034,132 @@ bool Score::appendScore(Score* score, bool addPageBreak, bool addSectionBreak)
+            else if (staff->currentTimeSigTick(ctick) == ctick &&
+                     (*staff->timeSig(ctick - Fraction::fromTicks(1))) == (*ostaff->timeSig(otick))) {
+                        Segment* ns = firstAppendedMeasure->first(SegmentType::TimeSig);
+                        if (ns) ns->remove(ns->element(trackIdx));
```
new line for body

JoshuaBonn1 commented on libmscore/score.cpp:
```diff
@@ -2067,59 +2034,132 @@ bool Score::appendScore(Score* score, bool addPageBreak, bool addSectionBreak)
+                     staff->clef(ctick - Fraction::fromTicks(1)) == ostaff->clef(otick)) {
+                        Segment* ns = firstAppendedMeasure->first(SegmentType::Clef);
+                        if (!ns) ns = firstAppendedMeasure->first(SegmentType::HeaderClef);
+                        if (ns) ns->remove(ns->element(trackIdx));
```
new line for bodies. For the record, I'm only going to mention this for bodies that have some complexity. I don't really mind if you put `break` or `continue` on the same line as a simple if statement.

JoshuaBonn1 commented on libmscore/score.cpp:
```diff
@@ -2067,59 +2034,132 @@ bool Score::appendScore(Score* score, bool addPageBreak, bool addSectionBreak)
+
+            // check if clef signature needs to be changed
+            if (!(ostaff->clef(otick) == staff->clef(ctick))) {
+                  undoChangeClef(staff,firstAppendedMeasure,ostaff->clef(otick));
```
spaces after comma

JoshuaBonn1 commented on libmscore/score.cpp:
```diff
@@ -2067,59 +2034,132 @@ bool Score::appendScore(Score* score, bool addPageBreak, bool addSectionBreak)
+      Measure* cm = firstAppendedMeasure, * pm = cm->prevMeasure();
+      if (pm->timesig() == cm->timesig() && 
+          pm->ticks() + cm->ticks() == cm->timesig()) {
+            qDebug() << "Detected a pre-measure that should be joined with previous measure at " << tickOfAppend;
```
unnecessary?

JoshuaBonn1 commented on libmscore/score.cpp:
```diff
@@ -2067,59 +2034,132 @@ bool Score::appendScore(Score* score, bool addPageBreak, bool addSectionBreak)
+      // check if section starts with apick-up measure to be merged with end of previous section
+      Measure* cm = firstAppendedMeasure, * pm = cm->prevMeasure();
+      if (pm->timesig() == cm->timesig() && 
+          pm->ticks() + cm->ticks() == cm->timesig()) {
```
Do you really need to separate this into two lines? You use equally long if statements in other spots. Feel free to explain your reasoning.

JoshuaBonn1 commented on libmscore/score.cpp:
```diff
@@ -2067,59 +2034,132 @@ bool Score::appendScore(Score* score, bool addPageBreak, bool addSectionBreak)
+      if (pm->timesig() == cm->timesig() && 
+          pm->ticks() + cm->ticks() == cm->timesig()) {
+            qDebug() << "Detected a pre-measure that should be joined with previous measure at " << tickOfAppend;
+            cmdJoinMeasure(pm,cm);
```
Space after comma

JoshuaBonn1 commented on libmscore/score.cpp:
```diff
@@ -2067,59 +2034,132 @@ bool Score::appendScore(Score* score, bool addPageBreak, bool addSectionBreak)
+                  }
+            }
+      else if (!last()->lineBreak() && !last()->pageBreak())
+            last()->undoSetBreak(true, LayoutBreak::Type::LINE);
```
Needs brackets around for the else if statement because the if statement is multilined

JoshuaBonn1 commented on libmscore/score.cpp:
```diff
@@ -2067,59 +2034,132 @@ bool Score::appendScore(Score* score, bool addPageBreak, bool addSectionBreak)
+            score->cmdConcertPitchChanged(styleB(Sid::concertPitch), true);
+
+      // clone the measures
+      appendMeasuresFromScore(score,Fraction(0,1),score->last()->endTick());
```
Spaces after commas

JoshuaBonn1 commented on libmscore/staff.cpp:
```diff
@@ -455,6 +468,25 @@ TimeSig* Staff::nextTimeSig(const Fraction& tick) const
+Fraction Staff::currentTimeSigTick(const Fraction& tick) const
+      {
+      if (timesigs.empty())
+            return Fraction(0,1);
```
space after comma

velochy replied:
@JoshuaBonn1 Fraction(0,1) seems to be an idiom and is written like that in many other places. 
74 matches for Fraction(0, 1) as opposed to 399 for Fraction(0,1)

JoshuaBonn1 replied:
In that case, use a space after a comma. The standard coding practice across all the code is much more prevalent than someone who forgot a space. Since there is a chance of making it Fraction::zero, I understand that it will probably be replaced. 

Definitely not merge stopping though.

velochy replied:
@JoshuaBonn1 Ok. If there are more substantial comments, I will fix this along with them.

JoshuaBonn1 commented on libmscore/staff.cpp:
```diff
@@ -455,6 +468,25 @@ TimeSig* Staff::nextTimeSig(const Fraction& tick) const
+            return Fraction(0,1);
+      auto i = timesigs.upper_bound(tick.ticks());
+      if (i == timesigs.begin())
+            return Fraction(0,1);
```
space after comma

JoshuaBonn1 commented on libmscore/unrollrepeats.cpp:
```diff
@@ -0,0 +1,207 @@
+                  first = false;
+                  }
+            else // just append this section from the original to the new score
+                  score->appendMeasuresFromScore(original, startTick, endTick);
```
brackets around body from else statement because if statement is multi lined

JoshuaBonn1 commented on libmscore/unrollrepeats.cpp:
```diff
@@ -0,0 +1,207 @@
+            }
+
+      /*score->setLayoutAll();
+      score->doLayout();*/ // scanElements in repeatMarkings may crash without!
```
unnecessary?

JoshuaBonn1 commented on libmscore/unrollrepeats.cpp:
```diff
@@ -0,0 +1,207 @@
+
+      // re-create excerpt parts
+      if (!excerpts.isEmpty())
+            createExcerpts(score,excerpts);
```
space after comma

JoshuaBonn1 commented on mscore/musescore.cpp:
```diff
@@ -7630,3 +7635,14 @@ QmlPluginEngine* MuseScore::getPluginEngine()
+      {
+      MasterScore * score = original->unrollRepeats();
+      setCurrentScoreView(appendScore(score));
+      }
```
Add new line (I believe that's the standard)

JoshuaBonn1 commented on mscore/shortcut.cpp:
```diff
@@ -3560,6 +3560,16 @@ Shortcut Shortcut::_sc[] = {
+         STATE_NORMAL,
+         "unroll-repeats",
+         QT_TRANSLATE_NOOP("action","Unroll Repeats"),
+         QT_TRANSLATE_NOOP("action","Unroll Repeats"),         
```
spaces after commas

velochy commented:
Resolved all comments left by @JoshuaBonn1 except for the Fraction(0,1) styling issue (reason given above).

dmitrio95 commented on libmscore/score.cpp:
```diff
@@ -1978,148 +1978,182 @@ void Score::removeAudio()
+//   appendScore
+//---------------------------------------------------------
+
+bool Score::appendScore(Score* score, bool addPageBreak, bool addSectionBreak)
```
Could this function perhaps be moved above the `appendMeasuresFromScore`? It looks like it remained unchanged after this PR (if I am correct), and the version control system doesn't understand that. If it is possible, it would be better if git could handle this correctly.

dmitrio95 commented on libmscore/staff.cpp:
```diff
@@ -327,6 +327,19 @@ Fraction Staff::nextClefTick(const Fraction& tick) const
+//    return 0, if no such clef
+//---------------------------------------------------------
+
+Fraction Staff::currentClefTick(const Fraction& tick) const
```
Does it make sense to pass fractions by reference? Fraction should be at most twice larger than a pointer (if both pointers and `int` have the same size, sometimes it isn't, like for g++ on my Debian system), and passing pointers make its dereferencing needed to operate on this fraction. Of course, references are not necessarily implemented as pointers but there doesn't seem to be much choice if compiler cannot inline the function.

Of course, the scope of this question is beyond this single PR.

velochy replied:
@dmitrio95 it seems to be done this way in most other places, so i just followed the general trend.

dmitrio95 replied:
Yes, this is a separate question, so I believe it is OK to leave it as is in this pull request. Still it would be interesting to see whether it makes any measurable performance difference, maybe I'll try to check it later.

dmitrio95 commented on mscore/musescore.h:
```diff
@@ -742,6 +742,8 @@ class MuseScore : public QMainWindow, public MuseScoreCore {
       bool exportPartsPdfsToJSON(const QString& inFilePath, const QString& outFilePath = "/dev/stdout");
       /////////////////////////////////////////////////
 
+      void scoreUnrolled(MasterScore * original);
```
Concerning the style, `Type* somename` is usually used in MuseScore:
https://github.com/musescore/MuseScore/blob/0c4aefdd6c5bf49eb7b92c36ffa7bd858296174f/mscore/musescore.h#L453

dmitrio95 commented on libmscore/unrollrepeats.cpp:
```diff
@@ -0,0 +1,196 @@
+
+namespace Ms {
+
+void removeRepeatMarkings(Score* score)
```
If these functions are not members of any class and they are used only within this file you may want to declare them `static`, like here:
https://github.com/musescore/MuseScore/blob/0c4aefdd6c5bf49eb7b92c36ffa7bd858296174f/libmscore/layout.cpp#L3169

Not strictly necessary, of course.

drowo commented:
What's the status here? Could this possibly be merged to master?

velochy commented:
As the author, I have implemented all recommendations and believe this to be ready for a merge.

drowo commented:
Ok, let't ask another way: is there a particular reason why this can't be merged to master right now?

jthistle commented:
A little bit of patience ;)
I'm sure it'll be merged in good time, like my PR was.

drowo commented:
I just want to make sure, this really useful tool is not forgotten :-)

anatoly-os commented:
I will test the PR and merge if everything is ok.

velochy commented:
@anatoly-os thank you. 
There are two automatic test cases which should cover most of the complex issues already, but there is always room for even more testing :)

drowo commented:
👍🎶😁

velochy commented:
@anatoly-os have you had a chance to review?

RobFog commented:
3.0.5 was only released yesterday, so probably not… :-)

drowo commented:
Any news here?

anatoly-os commented:
The code looks good. Let me check how it works...

anatoly-os commented:
I noticed a few things:

1. Unrolled score is created with empty name. I suggest using "ScoreName" + "_Unrolled".
2. The unrolled score has different layout which is ok, but looks interesting. Are different layout properties applied? @velochy Do you know the reason? See attached picture.

<img width="1515" alt="123 2019-03-18 15-56-07" src="https://user-images.githubusercontent.com/18065034/54535060-63374780-4996-11e9-9027-72ff2c1d5f20.png">

velochy commented:
@anatoly-os I do not know why layout changes. If you look at the code, it copies the score object, so I am not sure why the layout properties are not copied with it.

velochy commented:
@anatoly-os could you attach the file you had this layouting problem with? I ran a few files I had at hand and in them the spacing seemed to remain the same. 

Also - it would help if you had ideas regarding which settings might not get transferred in the score object cloning. Or, if it is not important in your opinion, just say so and I will get the other two things you mentioned fixed

anatoly-os commented:
@velochy as you can see it is just a default empty score. I added voltas to the 6th and 7th measures before running the unrolling.

This is not really important, because resulting score is different anyway. So, please address the two things I mentioned and we will merge the changes to master and let people test the unrolling there.

velochy commented:
@anatoly-os made the two changes you requested:
a) Moved it in the menu to below "Resequence rehersal marks"
b) Gave the new score a name in the form <original name>_unrolled

drowo commented:
@anatoly-os thanks for merging!
